File: Data Structures & Algorithms/is-anagram/submission-3.js
class Solution {
    /**
     * @param {string} s
     * @param {string} t
     * @return {boolean}
     */
    isAnagram(s, t) {
        const cache = new Map()
        for (const si of s) {
            if (cache.has(si)) {
                cache.set(si, cache.get(si) + 1)
            } else {
                cache.set(si, 1)
            }
        }

        for (const ti of t) {
            if (!cache.has(ti) || cache.get(ti) === 0) return false;
            cache.set(ti, cache.get(ti) - 1)
        }

        for (const [k, v] of cache) {
            if (v > 0) return false;
        }

        return true;
    }
}

File: Data Structures & Algorithms/course-schedule/submission-3.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        deps = {}
        for pre in prerequisites:
            [a, b] = pre
            if not a in deps:
                deps[a] = set()
            deps[a].add(b)
        
        cache = {}
        def is_looped(a: int, seen: set[int]):
            if a in cache:
                return cache[a]
            if a in seen:
                cache[a] = True
                return True
            if not a in deps:
                cache[a] = False
                return False

            for b in deps[a]:
                if is_looped(b, seen.union([a])):
                    cache[a] = True
                    return True
            cache[a] = False
            return False

        for a in deps:
            if is_looped(a, set()):
                return False
        return True